### PR TITLE
PreserveCompilationContext needs to be set only if a project has any …

### LINKED ISSUE #2077: PreserveCompilationContext needs to be set only if a project has any Razor files
[ ] Fix the sdk
[ ] Clean up Identity UI csproj

### PR DESCRIPTION
**Pull request by pranavkm**
…Razor files

Fixes #2077

### CODE REVIEW

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets [hunk 1]
```diff
@@ -186,6 +191,16 @@ Copyright (c) .NET Foundation. All rights reserved.
   <Target Name="RazorCompile" DependsOnTargets="$(RazorCompileDependsOn)">
   </Target>
 
+  <!--
+    PreserveCompilationContext needs to enabled by default only for "applications" (OutputType = exe) which have one or more Razor files.
+    We cannot only inspect ItemGroups globally, therefore setting this must be done inside a target. We wire this up to run before Build.
+  -->
+  <Target Name="_InitialzePreserveCompilationContext">
+    <PropertyGroup Condition="'$(PreserveCompilationContext)' == '' AND '$(OutputType)' == 'exe' AND '@(Content->AnyHaveMetadataValue('Extension', '.cshtml'))' == 'true'">
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: Well aren't you a fancy boy?

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets [hunk 2]
```diff
@@ -186,6 +191,16 @@ Copyright (c) .NET Foundation. All rights reserved.
   <Target Name="RazorCompile" DependsOnTargets="$(RazorCompileDependsOn)">
   </Target>
 
+  <!--
+    PreserveCompilationContext needs to enabled by default only for "applications" (OutputType = exe) which have one or more Razor files.
+    We cannot only inspect ItemGroups globally, therefore setting this must be done inside a target. We wire this up to run before Build.
+  -->
+  <Target Name="_InitialzePreserveCompilationContext">
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: This is misspelled yo

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets [hunk 3]
```diff
@@ -58,6 +58,11 @@ Copyright (c) .NET Foundation. All rights reserved.
       RazorCoreCompile
     </RazorCompileDependsOn>
 
+    <BuildDependsOn>
+      _InitialzePreserveCompilationContext;
+      $(BuildDependsOn)
+    </BuildDependsOn>
```
Comment (rynowak) [inline src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: Is there something more granular we can attach to?

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs [hunk 4]
```diff
@@ -25,25 +25,21 @@ public Task Build_SimpleMvc_UsingDesktopMSBuild_CanBuildSuccessfully()
 
         private async Task Build_SimpleMvc_CanBuildSuccessfully(MSBuildProcessKind msBuildProcessKind)
         {
-            var result = await DotnetMSBuild("Build", "/p:RazorCompileOnBuild=true", msBuildProcessKind: msBuildProcessKind);
+            var result = await DotnetMSBuild("Build", "/p:RazorCompileOnBuild=true /t:_IntrospectPreserveCompilationContext", msBuildProcessKind: msBuildProcessKind);
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs]: Let's not overload these tests. 

Can we instead add verification to this test that the compilation context is produced and then verify 1-2 negative cases (library, no razor files)?

File: src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets (hunk 3 above)
Comment (pranavkm) [reply src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Sdk.Razor.CurrentVersion.targets]: Might be able to wire this up to `PrepareForBuildDependsOn`.

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs (hunk 4 above)
Comment (pranavkm) [reply test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs]: Sure

**Comment by pranavkm** [comment]
🆙 📅

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs [hunk 5]
```diff
@@ -19,5 +19,37 @@ public class BuildIntrospectionTest : MSBuildIntegrationTestBase
...
         }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task RazorSdk_SetsPreserveCompilationContextForProjectsWithCshtmlFiles()
+        {
+            var result = await DotnetMSBuild("Build", "/t:_IntrospectPreserveCompilationContext");
+
+            Assert.BuildPassed(result);
+            Assert.BuildOutputContainsLine(result, "PreserveCompilationContext: true");
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs]: Is it possible to verify a build output rather than text?

Relying on introspection is fragile - if we set the property after its been used we wouldn't know it broke. We should use it as a last resort?

Comment (pranavkm) [reply test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs]: Sure. I can make sure the output file has the requisite `CompilationLibrary` instances Mvc would care about

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs [hunk 6]
```diff
@@ -19,5 +19,37 @@ public class BuildIntrospectionTest : MSBuildIntegrationTestBase
...
+
+            var result = await DotnetMSBuild("Build", "/t:_IntrospectPreserveCompilationContext");
+
+            Assert.BuildPassed(result);
+            Assert.BuildOutputContainsLine(result, "PreserveCompilationContext: false");
+        }
+
+        [Fact]
+        [InitializeTestProject("ClassLibrary")]
+        public async Task RazorSdk_DoesNotSetPreserveCompilationContextForClassLibraryProjects()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs]: Writing this test requires creating a true ClassLibrary. The changes to support this get us in a better place and it's not overtly intrusive IMO

Comment (rynowak) [reply test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs]: Sure

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs [hunk 7]
```diff
@@ -228,5 +219,64 @@ public Task Build_SimpleMvc_UsingDesktopMSBuild_CanBuildSuccessfully()
...
+            
+            var projectReference = dependencyContext.CompileLibraries.First(l => l.Name == "SimpleMvc");
+            Assert.NotEmpty(packageReference.Assemblies);
+            
+            Assert.Contains(customDefine, dependencyContext.CompilationOptions.Defines);
+        }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task Build_WithoutProducesDepsFileWithotCompiilationContext()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs]: `Build_WithViews_ProducesDepsFileWithCompilationContext`

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs [hunk 8]
```diff
@@ -228,5 +219,64 @@ public Task Build_SimpleMvc_UsingDesktopMSBuild_CanBuildSuccessfully()
             Assert.BuildOutputContainsLine(result, $@"CompileResource: {Path.Combine("Areas", "Products", "Pages", "_ViewStart.cshtml")} /Areas/Products/Pages/_ViewStart.cshtml");
             Assert.BuildOutputContainsLine(result, $@"CompileResource: {Path.Combine("..", "LinkedDir", "LinkedFile.cshtml")} /LinkedFileOut/LinkedFile.cshtml");
         }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task Build_WithViewsProducesDepsFileWithCompilationContext()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntegrationTest.cs]: `Build_WithViews_ProducesDepsFileWithCompilationContext`

File: test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs [hunk 9]
```diff
@@ -19,5 +19,37 @@ public class BuildIntrospectionTest : MSBuildIntegrationTestBase
             Assert.BuildOutputContainsLine(result, $"UpToDateCheckInput: {Path.Combine("Views", "Home", "Index.cshtml")}");
             Assert.BuildOutputContainsLine(result, $"UpToDateCheckInput: {Path.Combine("Views", "_ViewStart.cshtml")}");
         }
+
+        [Fact]
+        [InitializeTestProject("SimpleMvc")]
+        public async Task RazorSdk_SetsPreserveCompilationContextForProjectsWithCshtmlFiles()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs]: Do we need these introspection tests now that we have the 'real thing'? IMO they don't add much value and can be removed.

Comment (pranavkm) [reply test/Microsoft.AspNetCore.Razor.Design.Test/IntegrationTests/BuildIntrospectionTest.cs]: Yeah, I can take them out. Faster test runs
